cmd/cluster-agent/admissioncontroller/listener.go

```diff
@@ -0,0 +1,203 @@
+package admissioncontroller
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"k8s.io/api/admission/v1beta1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/kubernetes/pkg/apis/core/v1"
+
+	"github.com/DataDog/datadog-agent/cmd/agent/common/jsonpatch"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var (
+	// Scheme is the schemed used for validation and mutation of resources
+	scheme = runtime.NewScheme()
+
+	// Codecs is the codec used for en/de coding
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	corev1.AddToScheme(scheme)
+	admissionregistrationv1beta1.AddToScheme(scheme)
+	// defaulting with webhooks:
+	// https://github.com/kubernetes/kubernetes/issues/57982
+	v1.AddToScheme(scheme)
+}
+
+func status(w http.ResponseWriter, r *http.Request) {
+	w.Write([]byte("status OK"))
+}
+
+func serve(w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	log.Debug("admission controller request body: %v", body)
+
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+	responseAdmissionReview := v1beta1.AdmissionReview{}
+
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		responseAdmissionReview.Response = newAdmissionResponseWithError(err)
+	} else {
+		if requestedAdmissionReview.Request == nil {
+			w.WriteHeader(http.StatusBadRequest)
+			log.Info("received empty request")
+			return
+		}
+
+		responseAdmissionReview.Response = handleAdmissionReview(requestedAdmissionReview)
+	}
+
+	// Return the same UID
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+
+	log.Debug("admission controller response: %v", responseAdmissionReview.Response)
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		log.Error(err)
+		return
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		log.Error(err)
+		return
+	}
+}
+
+// Returns a simple response with the provided error.
+// The webhook is still considered as allowed, in order
+// to not interfere with the user's deployment.
+func newAdmissionResponseWithMessage(message string, params ...interface{}) *v1beta1.AdmissionResponse {
```

Comment:
```suggestion
func newResponseWithMessage(format string, a ...interface{}) *v1beta1.AdmissionResponse {
```
The argument names I picked from https://golang.org/pkg/fmt/#Sprintf which is exactly how they are used